system/camerad/cameras/camera_qcom2.cc

```diff
@@ -927,41 +927,50 @@ void CameraState::set_camera_exposure(float grey_frac) {
+  auto initCameraState = [type = buf.stream_type]() {
+    if (type == VISION_STREAM_DRIVER) return &cereal::Event::Builder::initDriverCameraState;
+    if (type == VISION_STREAM_ROAD) return &cereal::Event::Builder::initRoadCameraState;
+    return &cereal::Event::Builder::initWideRoadCameraState; }();
+
+  for (uint32_t cnt = 0; !do_exit; ++cnt) {
+    // Acquire the buffer; continue if acquisition fails
+    if (!buf.acquire()) continue;
+
+    MessageBuilder msg;
+    auto framed = (msg.initEvent().*initCameraState)();
+    fill_frame_data(framed, buf.cur_frame_data, this);
+
+    // Log raw frames for road camera
+    if (env_log_raw_frames && buf.stream_type == VISION_STREAM_ROAD && cnt % 100 == 5) {  // no overlap with qlog decimation
+      framed.setImage(get_raw_frame_image(&buf));
+    }
+    // Log frame id for road and wide road cameras
+    if (buf.stream_type != VISION_STREAM_DRIVER) {
+      LOGT(buf.cur_frame_data.frame_id, "%s: Image set", pub_name);
+    }
 
-void process_road_camera(MultiCameraState *s, CameraState *c, int cnt) {
-  const CameraBuf *b = &c->buf;
+    // Process camera registers and set camera exposure
+    ci->processRegisters(this, framed);
+    set_camera_exposure(set_exposure_target(&buf, ae_xywh, 2, buf.stream_type != VISION_STREAM_DRIVER ? 2 : 4));
 
-  MessageBuilder msg;
-  auto framed = c == &s->road_cam ? msg.initEvent().initRoadCameraState() : msg.initEvent().initWideRoadCameraState();
-  fill_frame_data(framed, b->cur_frame_data, c);
-  if (env_log_raw_frames && c == &s->road_cam && cnt % 100 == 5) {  // no overlap with qlog decimation
-    framed.setImage(get_raw_frame_image(b));
+    // Send the message
+    pm->send(pub_name, msg);
+    if (buf.stream_type == VISION_STREAM_ROAD && cnt % 100 == 3) {
+      publish_thumbnail(pm, &buf);  // this takes 10ms???
+    }
   }
-  LOGT(c->buf.cur_frame_data.frame_id, "%s: Image set", c == &s->road_cam ? "RoadCamera" : "WideRoadCamera");
-
-  c->ci->processRegisters(c, framed);
-  s->pm->send(c == &s->road_cam ? "roadCameraState" : "wideRoadCameraState", msg);
-
-  const int skip = 2;
-  c->set_camera_exposure(set_exposure_target(b, c->ae_xywh, skip, skip));
 }
 
 void cameras_run(MultiCameraState *s) {
   LOG("-- Starting threads");
   std::vector<std::thread> threads;
-  if (s->driver_cam.enabled) threads.push_back(start_process_thread(s, &s->driver_cam, process_driver_camera));
-  if (s->road_cam.enabled) threads.push_back(start_process_thread(s, &s->road_cam, process_road_camera));
-  if (s->wide_road_cam.enabled) threads.push_back(start_process_thread(s, &s->wide_road_cam, process_road_camera));
+  if (s->driver_cam.enabled) threads.emplace_back(&CameraState::run, &s->driver_cam, s->pm, "driverCameraState");
+  if (s->road_cam.enabled) threads.emplace_back(&CameraState::run, &s->road_cam, s->pm, "roadCameraState");
+  if (s->wide_road_cam.enabled) threads.emplace_back(&CameraState::run, &s->wide_road_cam, s->pm, "wideRoadCameraState");
```

Comment:
would be great to refactor away from these if's too

Reply:
This might make this PR too large. There's still a lot of work to be done to clean up the camerad code, which will require a series of PRs. As a start, I have submitted a PR to add a CameraConfig struct for initializing CameraState in the constructor. https://github.com/commaai/openpilot/pull/33034